cpp/src/dictionary/dictionary_factories.cu

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/gather.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+
+namespace cudf
+{
+namespace detail
+{
+namespace
+{
+
+// Sort functor for column row positions.
+template <bool has_nulls = true>
+struct sort_functor
+{
+    column_device_view d_column;
+    __host__ __device__ bool operator()(size_type lhs_index, size_type rhs_index)
+    {
+        auto comparator = experimental::element_relational_comparator<has_nulls>{
+                          d_column, d_column, null_order::AFTER }; // put nulls at the end
+        auto result = experimental::type_dispatcher(d_column.type(), comparator,
+                                                    lhs_index, rhs_index);
+        return result == experimental::weak_ordering::LESS; // always sort ascending
+    }
+};
+
+// Copy functor used for locating and marking unique values.
+template <bool has_nulls = true>
+struct copy_unique_functor
+{
+    column_device_view d_column;
+    const int32_t* d_ordinals;
+    int32_t* d_indices;
+
+    __device__ bool operator()(int32_t idx)
+    {
+        if( idx==0 )
+        {   // first element is always unique
+            d_indices[0] = 0;
+            return true;
+        }
+        // check if adjacent elements match
+        auto lhs_index = d_ordinals[idx-1];
+        auto rhs_index = d_ordinals[idx];
+        auto result = !cudf::experimental::type_dispatcher( d_column.type(),
```

Comment:
> This is doing a type dispatch in the inner loop, which is expensive and unnecessary. You only need to dispatch the type once for the entire column.

This is the same pattern used by `row_equality_comparator` here:
https://github.com/rapidsai/cudf/blob/db1dc2d18f4af4dd9fa24df30d375838e95bbb8d/cpp/include/cudf/table/row_operators.cuh#L207

Reply:
That's because `row_equality_comparator` has to work on rows, which necessitates device code dispatch, once per column in the row. In this case, you're operating on a single column, so you only need to do the type dispatch once _before_ the kernel.